src/adapter/src/coord/message_handler.rs

```diff
@@ -261,12 +261,56 @@ impl<S: Append + 'static> Coordinator<S> {
             return;
         }
 
+        //TODO(petrosagg): Figure out a way to execute the subsources and sources atomically. With
+        //                 the way it's written now it is possible for subsources to be created but
+        //                 not the source object.
+        // First plan all subsource statements
+        let mut subsource_plans = vec![];
+        for stmt in subsource_stmts {
+            let catalog = self.catalog.for_session(&session);
+
+            let (stmt, depends_on) = match mz_sql::names::resolve(&catalog, stmt) {
+                Ok(ok) => ok,
+                Err(e) => return tx.send(Err(e.into()), session),
+            };
+
+            let plan =
+                match self.plan_statement(&mut session, Statement::CreateSource(stmt), &params) {
+                    Ok(Plan::CreateSource(plan)) => plan,
+                    Ok(_) => {
+                        unreachable!("planning CREATE SOURCE must result in a Plan::CreateSource")
+                    }
+                    Err(e) => return tx.send(Err(e), session),
+                };
+            let depends_on = depends_on.into_iter().collect();
+            subsource_plans.push((plan, depends_on));
+        }
+
+        // Sequence all the subsources
+        for (plan, depends_on) in subsource_plans {
+            if let Err(e) = self
+                .sequence_create_source(&mut session, plan, depends_on)
+                .await
+            {
+                tx.send(Err(e), session);
+                return;
+            }
+        }
```

Comment:
As you mention above, calling `sequence_create_source` multiple times in the same transaction is a bit scary and will break any atomic guarantees.

You could update `sequence_create_source` (or create a new method) so that it takes a list of `plan`s and `depends_on`s. Then within the method you would run `catalog_transact` with multiple `catalog::Op::CreateItem`, and then finally create multiple collections.

Reply:
Unfortunately this isn't possible (I think) because the final statement references objects created by the first statements. So I can't come up with a plan because I can't come up with a name resolved statement until the catalog has been updated. Is there a way to work around that?

Reply:
hmm, I see what you mean. Part of `catalog_transact` is to create a copy of the catalog and update this copy one operation at a time. So it should in theory be possible to process one source at a time and then resolve the next source based on the updates to the catalog of the previous sources. However, I don't think the current set up of `catalog_transact` makes this easy or even possible. 

The problem here though is that if we crash after executing some `sequence_create_source` but not others, then we'll re-boot into a partial state and I'm not sure what will happen.

Reply:
> The problem here though is that if we crash after executing some sequence_create_source but not others, then we'll re-boot into a partial state and I'm not sure what will happen.

Yeah, definitely need to find a way to fix this.

You could probably rig something up with `ConnCatalog`. Add a field like `proposed_entries: HashMap<GlobalId, CatalogEntry>`, and teach `ConnCatalog.resolve_entry` to look in `proposed_entries` before the real catalog. Then you could build up `proposed_entries` as you go, and then commit all at once.

I'm not actually looking at the code, so I might be off on the details. But I think the general shape of this solution would work.

Reply:
I've been experimenting a little bit with this: https://github.com/MaterializeInc/materialize/pull/15030/files. It changes `Catalog::transact` so that instead of taking two passes over each `Op`, it only takes a single pass. That way each `Op` has an updated catalog from the results of the previous `Op`. If that change works, then I think it should be possible to combine all these create sources into a single catalog transaction.

Reply:
`Catalog::transact` already makes an entire copy of the state during a transaction, which is essentially the purposed entries. So if we also add another field called `proposed_entries`, we may end up duplicating a lot of logic or causing confusion.

Reply:
@jkosh44 I was also working on an alternative solution that I just published in this PR. All the `Either<Aug, Raw>` stuff is gone. Instead of that I changed purification so that it can output multiple createsource statements, each associated with a **transient** global id. This allows purification to create as many objects as it wants and define interrelationships using those transient ids.

When purification is done the coordinator will go through all the returned object and allocate a global id for each of them and then walk the statements and any time it finds a reference to a previously transient id it replaces it with the global id that was just minted. So far nothing has touched the catalog.

Now we can plan all the statements in one go and also sequence them in one go with a single catalog transact call. What do you think?

Reply:
Actually, reading your comment above it sounds like it is required to do your change too. I didn't know that catalog_transact didn't support interelated Ops in a single transact operation

Reply:
Well `catalog_transact` basically does the following in order (leaving out some unnecessary details):

1. Start a new stash transaction.
2. For each operation, update the stash transaction (and do some other misc tasks like generate writes to system tables).
3. Make a copy of the in-memory catalog contents.
4. For each operation update the copy of the catalog.
5. Commit stash transaction.
6. Commit the new catalog contents to memory.

So if step 2 for any operation depends on step 3 for some other operation, then you can't have interrelated Ops. However, if step 2 does not depend on step 3 then you should be fine. I'm actually not sure if that would be the case here.

Reply:
I took a look, I think `CREATE SOURCE` will probably fail in step 2, due to a recent change. During step 2 for `CREATE SOURCE` we check to see if the source has unstable uses:
https://github.com/MaterializeInc/materialize/blob/91bb0c84cede72a7f2e55da339e23b38b191b8a8/src/adapter/src/catalog.rs#L3609-L3615
To do that we check and see if any of the source's dependencies are unstable:
https://github.com/MaterializeInc/materialize/blob/91bb0c84cede72a7f2e55da339e23b38b191b8a8/src/adapter/src/catalog.rs#L4266-L4272
To do that we look up all of the dependencies in the catalog:
https://github.com/MaterializeInc/materialize/blob/91bb0c84cede72a7f2e55da339e23b38b191b8a8/src/adapter/src/catalog.rs#L264-L265
If the dependency doesn't exist then we panic:
https://github.com/MaterializeInc/materialize/blob/91bb0c84cede72a7f2e55da339e23b38b191b8a8/src/adapter/src/catalog.rs#L307-L309

Reply:
Finally got my PR to pass CI: https://github.com/MaterializeInc/materialize/pull/15030. With that change your approach should be fine. Feel free to still take a hacky approach in the meantime if it doesn't get merged soon.